code/functions.py

```diff
@@ -0,0 +1,165 @@
+from holoviews import plotting
+from holoviews import Options, Store
+import pfaffian as pf
+import kwant
+from types import SimpleNamespace
+
+
+kron_prods = ['s0s0', 's0sx', 's0sy', 's0sz',
+              'sxs0', 'sxsx', 'sxsy', 'sxsz',
+              'sys0', 'sysx', 'sysy', 'sysz',
+              'szs0', 'szsx', 'szsy', 'szsz']
+
+__all__ = ['sigma0', 'sigmax', 'sigmay', 'sigmaz', 'spectrum',
+           'plot_bands', 'h_k', 'plot_bands', 'find_pfaffian',
+           'kitaev_chain', 'spinful_kitaev_chain', 'nanowire_chain'] + kron_prods
+
+options = Store.options(backend='matplotlib')
+options.Contours = Options('style', linewidth=2, color='k')
+options.Contours = Options('plot', aspect='square')
+options.HLine = Options('style', linestyle='--', color='b', linewidth=2)
+options.VLine = Options('style', linestyle='--', color='r', linewidth=2)
+options.Image = Options('style', cmap='gist_heat_r')
+options.Image = Options('plot', title_format='{label}')
+options.Path = Options('style', linewidth=1.2, color='k')
+options.Path = Options('plot', aspect='square', title_format='{label}')
+options.Curve = Options('style', linewidth=2, color='k')
+options.Curve = Options('plot', aspect='square', title_format='{label}')
+options.Overlay = Options('plot', show_legend=False, title_format='{label}')
+
+sigma0 = np.array([[1., 0.], [0., 1.]])
+sigmax = np.array([[0., 1.], [1., 0.]])
+sigmay = np.array([[0., -1j], [1j, 0.]])
+sigmaz = np.array([[1., 0.], [0., -1.]])
+
+s0s0 = np.kron(sigma0, sigma0)
+s0sx = np.kron(sigma0, sigmax)
+s0sy = np.kron(sigma0, sigmay)
+s0sz = np.kron(sigma0, sigmaz)
+sxs0 = np.kron(sigmax, sigma0)
+sxsx = np.kron(sigmax, sigmax)
+sxsy = np.kron(sigmax, sigmay)
+sxsz = np.kron(sigmax, sigmaz)
+sys0 = np.kron(sigmay, sigma0)
+sysx = np.kron(sigmay, sigmax)
+sysy = np.kron(sigmay, sigmay)
+sysz = np.kron(sigmay, sigmaz)
+szs0 = np.kron(sigmaz, sigma0)
+szsx = np.kron(sigmaz, sigmax)
+szsy = np.kron(sigmaz, sigmay)
+szsz = np.kron(sigmaz, sigmaz)
+
+
+def spectrum(sys, xticks, yticks, xdim, ydim, xlims=None, ylims=None, **kwargs):
+    p = SimpleNamespace(**kwargs)
+    
+    for key, value in kwargs.items():
+        try:
+            if len(value) > 0:
+                changing_variable = key
+                changing_values = value
+        except:
+            TypeError
+    
+    def energy(x):
+        p.__dict__[changing_variable] = x
+        H = sys.hamiltonian_submatrix(args=[p])
+        return np.linalg.eigvalsh(H)
+    
+    spectrum = np.array([energy(x) for x in changing_values])
+    
+    plot = hv.Path((changing_values, spectrum), kdims=[xdim, ydim])
+    
+    if xlims is None:
+        xlims = slice(xticks[0], xticks[-1])
+    else:
+        xlims = slice(xlims[0], xlims[1])
+    if ylims is None:
+        ylims = slice(yticks[0], yticks[-1], None)
+    else:
+        ylims = slice(ylims[0], ylims[1])
+    
+    return plot[xlims, ylims](plot={'xticks':xticks, 'yticks':yticks})
+
+def h_k(sys, p, phase_fac=1):
```

Comment:
Please copy the function body from `kwant.physics.dispersion.Bands`: your current version won't work if not all hoppings are present. Specifically I mean the reshaping of the hopping done [here](https://gitlab.kwant-project.org/kwant/kwant/blob/master/kwant/physics/dispersion.py#L50)